pkg/hubble/parser/threefour/parser.go

```diff
@@ -664,9 +664,12 @@ func (p *Parser) decodeNetworkInterface(tn *monitor.TraceNotify, dbg *monitor.De
 		return nil
 	}
 
-	// if the interface is not found, `name` will be an empty string and thus
-	// omitted in the protobuf message
-	name, _ := p.linkGetter.GetIfNameCached(int(ifIndex))
+	var name string
+	if p.linkGetter != nil {
```

Comment:
Is it possible to avoid nil checks like this per message?

`p.linkGetter` should be *guaranteed* to be non-nil since creation. If someone creates a `Parser` without using the appropriate constructors, then it's their problem.

There's two things I'm not fond of here: One, we're needlessly adding additional conditional checks for every message. Maybe that's mostly optimized out since branch prediction will figure that this field is always non-nil, but the check is still unnecessary. Two, now that the code checks whether `p.linkGetter` is non-nil, we force the developer / reader to consider that case as though it were a legitimate path through the program. Whereas from what I understand, we never expect this to be the case, so no-one should have to care about this case while reading/vetting the code.

Reply:
> Is it possible to avoid nil checks like this per message?
> 
> `p.linkGetter` should be _guaranteed_ to be non-nil since creation. If someone creates a `Parser` without using the appropriate constructors, then it's their problem.

In the failing fuzzer test, the `Parser` was created using the constructor, but passing `nil` for all getters: https://github.com/cncf/cncf-fuzzing/blob/ff428349a1b6f2c382c84fbbb3812f680c25e94b/projects/cilium/hubble_parser_fuzzer.go#L41 I don't see any code in the constructor that would guarantee the `linkGetter` (or any other getter) to be non-`nil` after construction.

> There's two things I'm not fond of here: One, we're needlessly adding additional conditional checks for every message. Maybe that's mostly optimized out since branch prediction will figure that this field is always non-nil, but the check is still unnecessary. Two, now that the code checks whether `p.linkGetter` is non-nil, we force the developer / reader to consider that case as though it were a legitimate path through the program. Whereas from what I understand, we never expect this to be the case, so no-one should have to care about this case while reading/vetting the code.

Agree on the needlessness of performing the check on every message. I followed this approach (i.e. checking the getter before access) because this is what is currently also done for other existing getters, e.g. `Parser.serviceGetter`.

One approach I see to avoid the `nil` check per message would be to check all getters being non-`nil` in the constructor and return an error otherwise. I'd prefer to do that for all getters at once, though. So I'd probably defer this to a follow-up PR so we can keep the fix for the failing fuzz test separate from refactoring the check.

Reply:
OK thanks.

Maybe another 2c, I think this may be highlighting a little bit of perhaps "non-idiomatic" go usage we have in the constructors here: When a fuzzer developer wants to integrate with this code in order to validate the parsing pieces, ideally they should be able to do this without a bunch of boilerplate for each of these getters. Perhaps if we design this interface a bit better, it would use noopgetters for these features unless the caller specifies them. Then, for the purposes of testing the intended target of the fuzzing, the relevant code would be included and all of these getters would be abstracted out (and hence could be ignored).

But yeah fair enough to follow up on this separately.